### ISSUE TITLE
Cannot send encryption packet. Falling back to cracked login

### ISSUE THREAD

**Issue by kristupas098** [games647/FastLogin#412]
[//]: # (Lines in this format are considered as comments and will not be displayed.)
[//]: # (Before reporting make sure you're running the **latest build** of the plugin and checked for existing issues!)

### What behaviour is observed:
[//]: # (What happened?)
In version 1.16.4 then you login to server it asks your password.
Plugin uses cracked player info instead of premium user

### What behaviour is expected:
[//]: # (What did you expect?)
To login without password and using premium account

### Steps/models to reproduce:
[//]: # (The actions that cause the issue. Please explain it in detail)
Only need to update server software (spigot,papermc) to minecraft version 1.16.4

### Screenshots (if applicable)
[//]: # (You can drop the files here directly)

### Plugin list:
[//]: # (This can be found by running `/pl`)
Authme, chatco, customhelpmessage, fastchunkpregenerator, fastlogin, infobook ,joinmessage, kiss-tpa, luckperms, nocheatplus ,oldschoolkill ,protocollib, randomspawn, realping, rules, serverlistplus, skinsrestorer

### Environment description
[//]: # (Server software with exact version number, Minecraft version, SQLite/MySQL/MariaDB, ...)
Minecraft 1.16.4, papermc git-256

### Plugin version or build number (don't write latest):
[//]: # (This can be found by running `/version plugin-name`.)
1.11-snapshot-d034e71

### Server Log:
[//]: # (No images please - only the textual representation)
[Hastebin](https://hastebin.com/) / [Gist](https://gist.github.com/) link of the error, stacktrace or the complete log (if any)
https://hastebin.com/keqemahodi.yaml

### Configuration:
[//]: # (No images please - only the textual representation)
[//]: # (remember to delete any sensitive data)
[Hastebin](https://hastebin.com/) / [Gist](https://gist.github.com/) link of your config.yml file 
https://hastebin.com/ozacitojuz.yaml

**Comment by gavenda** [comment]
Related, 1.16.4 changed the parameters for enabling encryption in https://github.com/games647/FastLogin/blob/master/bukkit/src/main/java/com/github/games647/fastlogin/bukkit/listener/protocollib/VerifyResponseTask.java#L173

```java
private boolean enableEncryption(SecretKey loginKey) throws IllegalArgumentException {
    try {
        //get the NMS connection handle of this player
        Object networkManager = getNetworkManager();

        //try to detect the method by parameters
        Method encryptMethod = FuzzyReflection
                .fromObject(networkManager).getMethodByParameters("a", SecretKey.class);

        //encrypt/decrypt following packets
        //the client expects this behaviour
        encryptMethod.invoke(networkManager, loginKey);
    } catch (Exception ex) {
        disconnect("error-kick", false, "Couldn't enable encryption", ex);
        return false;
    }

    return true;
}
```
Should be like this in 1.16.4
```java
private boolean enableEncryption(SecretKey loginKey) throws IllegalArgumentException {
    try {
        //get the NMS connection handle of this player
        Object networkManager = getNetworkManager();

        //try to detect the method by parameters
        Method encryptMethod = FuzzyReflection
                .fromObject(networkManager).getMethodByParameters("a", Cipher.class, Cipher.class);

        Cipher decryptCipher = Cipher.getInstance("AES/CFB8/NoPadding");
        decryptCipher.init(Cipher.DECRYPT_MODE, loginKey, new IvParameterSpec(loginKey.encoded))
        Cipher encryptCipher = Cipher.getInstance("AES/CFB8/NoPadding");
        encryptCipher.init(Cipher.ENCRYPT_MODE, loginKey, new IvParameterSpec(loginKey.encoded))

        //encrypt/decrypt following packets
        //the client expects this behaviour
        encryptMethod.invoke(networkManager, decryptCipher, encryptCipher);
    } catch (Exception ex) {
        disconnect("error-kick", false, "Couldn't enable encryption", ex);
        return false;
    }

    return true;
}
```

Could move to another issue if inappropriate

**Comment by thomaboehi** [comment]
```
[17:07:04 ERROR]: [FastLogin] Couldn't enable encryption
java.lang.IllegalArgumentException: Unable to find a in net.minecraft.server.v1_16_R3.NetworkManager
        at com.comphenix.protocol.reflect.FuzzyReflection.getMethodByParameters(FuzzyReflection.java:261) ~[?:?]
        at com.github.games647.fastlogin.bukkit.listener.protocollib.VerifyResponseTask.enableEncryption(VerifyResponseTask.java:180) ~[?:?]
        at com.github.games647.fastlogin.bukkit.listener.protocollib.VerifyResponseTask.verifyResponse(VerifyResponseTask.java:87) ~[?:?]
        at com.github.games647.fastlogin.bukkit.listener.protocollib.VerifyResponseTask.run(VerifyResponseTask.java:63) ~[?:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.run(Unknown Source) ~[?:1.8.0_241]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:1.8.0_241]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:1.8.0_241]
        at java.lang.Thread.run(Unknown Source) [?:1.8.0_241]
```

PaperMC 261
FastLogin Build #957

**Comment by Kamilkime** [comment]
Well... My quick fix for that (in the VerifyResponseTask class)  looks like this:
```java
private boolean enableEncryption(SecretKey loginKey) throws IllegalArgumentException {
    try {
        //get the NMS connection handle of this player
        Object networkManager = getNetworkManager();

        //try to detect the method by parameters
        Method encryptMethod = FuzzyReflection.fromObject(networkManager).getMethodByParameters("a", Cipher.class, Cipher.class);

        final Class<?> encryptionClass = Class.forName("net.minecraft.server.v1_16_R3.MinecraftEncryption");
        final Method aMethod = encryptionClass.getMethod("a", int.class, Key.class);

        final Object cipher = aMethod.invoke(null, 2, loginKey);
        final Object cipher1 = aMethod.invoke(null, 1, loginKey);

        //encrypt/decrypt following packets
        //the client expects this behaviour
        encryptMethod.invoke(networkManager, cipher, cipher1);
    } catch (Exception ex) {
        disconnect("error-kick", false, "Couldn't enable encryption", ex);
        return false;
    }

    return true;
}
```

It only works on 1.16.4, so if anyone needs that fast and can't wait for an official fix (like me) - you can get it here, it's only that one thing fixed:
https://drive.google.com/file/d/1zBkCCR2lqxIfCWSWRnHi4Qd1csbJ6IL3/view?usp=sharing

_I hope you don't mind that @games647_

**Comment by awgybop1** [comment]
@Kamilkime The plugin is distributed under the MIT License, so you're free to edit and distribute it with no problems!

Thanks for compiling the plugin with that fix - it works perfectly.

**Comment by amalon** [comment]
For the benefit of anybody else building from source, these are the additional imports needed with @Kamilkime's fix:

```java
import java.security.Key;
import javax.crypto.Cipher;

```

**Comment by cicatrice74** [comment]
Will this fix be injected in future dev builds ?

**Comment by Vidhin05** [comment]
Has this been fixed in #992, I'm still getting this error.
